go/contacts/contacts.go

```diff
@@ -0,0 +1,186 @@
+	FillUsernames(libkb.MetaContext, []keybase1.ProcessedContact)
+}
+
+// ResolveContacts resolves contacts with cache for UI. See API documentation
+// in phone_numbers.avdl
+//
+// regionCode is optional, user region should be provided if it's known. It's
+// used when resolving local phone numbers, they are assumed to be local to the
+// user, so in the same region.
+func ResolveContacts(mctx libkb.MetaContext, provider ContactsProvider, contacts []keybase1.Contact,
+	regionCode keybase1.RegionCode) (res []keybase1.ProcessedContact, err error) {
+
+	if len(contacts) == 0 {
+		mctx.Debug("`contacts` is empty, nothing to resolve")
+		return res, nil
+	}
+
+	// Collect phone numbers and emails from Contact list. Iterate through
+	// contacts and descend into components. We will be passing phone number
+	// list and email list to the provider to try to resolve them.
+	type contactRef struct {
+		// Use this struct to point back from phoneNumbers or emails entry to
+		// our contacts list. We need a way to associate result from the
+		// provider to the contact and component they come from.
+		contactIndex   int
+		componentIndex int
+	}
+	var phoneNumbers []keybase1.RawPhoneNumber
+	var phoneComps []contactRef
+	var emails []keybase1.EmailAddress
+	var emailComps []contactRef
+	for contactI, k := range contacts {
+		for compI, component := range k.Components {
+			if component.Email != nil {
+				emails = append(emails, *component.Email)
+				emailComps = append(emailComps, contactRef{
+					contactIndex:   contactI,
+					componentIndex: compI,
+				})
+			}
+			if component.PhoneNumber != nil {
+				phoneNumbers = append(phoneNumbers, *component.PhoneNumber)
+				phoneComps = append(phoneComps, contactRef{
+					contactIndex:   contactI,
+					componentIndex: compI,
+				})
+			}
+		}
+	}
+
+	mctx.Debug("Going to look up %d emails and %d phone numbers", len(emails), len(phoneNumbers))
+
+	// contactIndex -> true for all contacts that have at least one compoonent resolved.
+	contactsFound := make(map[int]struct{})
+	usersFound := make(map[keybase1.UID]struct{})
+
+	insertResult := func(lookupRes ContactLookupResult, toContact contactRef) {
+		contactsFound[toContact.contactIndex] = struct{}{}
+
+		if _, found := usersFound[lookupRes.UID]; found {
+			// This user was already resolved by looking up another
+			// component or another contact.
+			return
+		}
+		contact := contacts[toContact.contactIndex]
+		component := contact.Components[toContact.componentIndex]
+
+		usersFound[lookupRes.UID] = struct{}{}
+
+		res = append(res, keybase1.ProcessedContact{
+			ContactIndex: toContact.contactIndex,
+			ContactName:  contact.Name,
+			Component:    component,
+			Resolved:     true,
+			Uid:          lookupRes.UID,
+			Following:    true, // assume following=true for now because this creates better display label.
+
+			// following, username (TODO???), and full name are filled later.
```

Comment:
we shouldn't fill in `following` from the server right? the client should run identify on the uid?

Reply:
yes, but I'm debating whether it should be client looking up its following list somehow or whether API sever should return this info in the `LookupBulk*` endpoints

Reply:
oh, I don't think we need to ID here... this is all server trust on first use anyway